### PR TITLE
[flang] Centralize automatic deallocation code in lowering

### PR DESCRIPTION
**Pull request by jeanPerier**
There are currently several places that automatically deallocate allocatble if they are allocated:
 - INTENT(OUT) allocatable are deallocated on entry in the callee
 - INTENT(OUT) allocatable are also deallocated on the caller side of BIND(C) function in case the implementation is in C.
 - Results of function returning allocatable are deallocated after usage.
 - OPENMP privatized allocatable are deallocated at the end of OPENMP region.

Introduce genDeallocateIfAllocated that centralize all this code, except for the function return that use genFreememIfAllocated since finalization is done separately currently.

`fir::factory::genFinalization` and `fir::factory::genInlinedDeallocation` are removed and replaced by genFreemem since their name were misleading: finalization was not called.

There is a fallout in the tests because previous generated code did not check the allocated status when doing inline deallocation. This was OK since free(null) is guaranteed to be a no-op, but this makes compiler code more complex, is a bit surprising in the generated IR IMHO, and it relied on knowing when genDeallocateBox inserts runtime calls or uses inlined code.

@kiranchandramohan, the OpenMP test in Lower/OpenMP/FIR/parallel-private-clause.f90 is updated because it contained hardcoded SSA value numbers that got broken by slight changes in the non OMP parts. There is no functional change in the OMP part.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Thanks @jeanPerier for the cleanup.

File: flang/include/flang/Lower/Allocatable.h [hunk 1]
```diff
@@ -57,6 +57,11 @@ void genDeallocateBox(AbstractConverter &converter,
                       const fir::MutableBoxValue &box, mlir::Location loc,
                       mlir::Value declaredTypeDesc = {});
 
+/// Deallocate an allocatable if it allocated at the end of its lifetime.
```
Comment (vzakhari) [inline flang/include/flang/Lower/Allocatable.h:60]: nit:
```suggestion
/// Deallocate an allocatable if it is allocated at the end of its lifetime.
```

File: flang/lib/Lower/ConvertCall.cpp [hunk 2]
```diff
@@ -461,10 +461,13 @@ fir::ExtendedValue Fortran::lower::genCallOpAndResult(
     allocatedResult->match(
         [&](const fir::MutableBoxValue &box) {
           if (box.isAllocatable() && !cleanupWithDestroy) {
-            // 9.7.3.2 point 4. Finalize allocatables.
+            // 9.7.3.2 point 4. Deallocate allocatable results. Note that
+            // finalization was done independently by calling
+            // genDerivedTypeDestroy above and is not triggered by this inline
+            // dealloction.
```
Comment (vzakhari) [inline flang/lib/Lower/ConvertCall.cpp:467]: typo
```suggestion
            // deallocation.
```

**Review by vzakhari — APPROVED** [review]
Looks good. Thank you, Jean!

**Pull request merged**